1-Docker

dahachm opened the pull request:
@konstantin009 @Shlepanets

**Docker containers with hadoop**

https://hub.docker.com/r/dahachm/hadoop

**Headnode** (namenode):
Get image:
```
$ docker pull dahachm/hadoop:headnode
```
Start container:
```
$ docker run --name headnode -dP -p 8088:8088 dahachm/hadoop:headnode
```

Access ResourceManager WebUI at *port 8080* on your host machine.


**Worker** (slave node):
Get container:
```
$ docker pull dahachm/hadoop:headnode
```
Start container:
```
$ docker run --name worker -dP -p 8042:8042 dahachm/hadoop:worker
```

Access NodeManager WebUI at *port 8042* on your host machine.

**Docker-compose file**

**docker-compose.yml** - for building up these containers and start it properly.

 Start composing:
```
$ docker-compose -f docker-compose.yml up -d 
```

dahachm committed:
Initial commit

**headnode_dockerfile** - dockerfile for building **headnode** image

**worker_dockerfile** - dockerfile for building **worker** image

**scripts** contains shell scripts:
    - **start-headnode.sh**:
            first to run when **headnode** container is starting.
            Perform formatting new cluster FS (by calling *hdfs namenode -format*).
            Then start *HDFS NameNode* and *YARN ResourceManager*.
    - **start-worker.sh**:
            first to run when **worker** container is starting.
            Start *HDFS DataNode* and *YARN NodeManager*.
    - **set-hadoop-conf.sh**:
            Load Hadoop config files and Set required parameters::
                    in *hadoop-env.sh* set JAVA_HOME, HADOOP_HOME, HADOOP_HEAPSIZE_MAX
                    in *core-site.xml* set HDFS_NAMENODE_HOSTNAME
                    in *hdfs-site.xml* set NAMENODE_DIRS and DATANODE_DIRS
                    in *yarn-site.xml* set YARN_RESOURCE_MANAGER_HOSTNAME, NODE_MANAGER_LOCAL_DIR, NODE_MANAGER_LOG_DIR

dahachm committed:
Update README.md

dahachm committed:
Update headnode_dockerfile

dahachm committed:
Update worker_dockerfile

dahachm committed:
Update README.md

dahachm committed:
**Update headnode_dockerfile**

	Add new script make-formatting.sh
		start formatting FS for new cluster

	Update start-headnode.sh
		delete calling FS formatting in order to
		prevent FS formatting on each start of the container

	Update headnode_dockerfile
		add new layer with running *make-formatting.sh* script

dahachm committed:
Update README.md

dahachm committed:
Add exec rights to make-formatting.sh script

dahachm committed:
Update adding hosts

        Add set-hosts-headnode.sh and set-hosts-worker.sh
                these scripts check current IP address of container (that was
                acquired from docker) and figure out the IP address of another node
                (headnode is first to start, so the IP addr of worker must be the
                next one in the same network).
                These scrtips start from start-headnode.sh and start-worker.sh.

        Update dockerfile for headnode and worker
                add calling set-hosts-headnode.sh and set-hosts-worker.sh
                at the beginning.

dahachm committed:
Update README.md

dahachm committed:
Update docker-compose.yml
	Add 'tty:true' to each services, in order to make
	container stay up after start

Shlepanets commented on 1-Docker/docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+
+services:
+  headnode:
+    image: dahachm/hadoop:headnode
```
Tags should be used to track versions of one image (ex. [Jenkins DockerHub](https://hub.docker.com/_/jenkins?tab=tags&page=1&ordering=last_updated) ), not to distinguish between headnode and worker.
There should be two separate images, i.e.: dahachm/hadoop_headnode:1.2 (or latest) and ahachm/hadoop_worker:1.15 (or latest)

dahachm replied:
Okay, was guessing is it right to put them in the same repo with different tags

The reason why I put them together was about the fact that thay are in the same task and they both are hadoop containers

Should I move one to another repo?

Shlepanets replied:
You can do it, if you'll have spare time, but this is not necessary to pass hometask =)

Shlepanets commented on 1-Docker/headnode_dockerfile:
```diff
@@ -0,0 +1,41 @@
+RUN /usr/local/set-hadoop-conf.sh
+
+RUN chown root:hadoop $HADOOP_HOME && \
+    chown root:hadoop $HADOOP_HOME/* && \
```
what is the expected result after that command?

dahachm replied:
*/usr/local/set-hadoop-conf.sh* - this script 
downloads hadoop config files (hadoop-env.sh, core-site.xml, hdfs-site.xml, yarn-site.xml) - the same that we had on Linux Exam - and set required parameters (JAVA_HOME, HADOOP_HOME, hostnames and dirs)

As we untared hadoop archive under root, *I needed to change the owning group of the hadoop working files and directories (to escape Permission denied fault)*

So in these lines I change the owning user and group of $HADOOP_HOME directory recursively (when I tested these commands, somehow **-r** parameter didn't work, so I put it in two lines)

Shlepanets replied:
1. Why didn't "chown -r" work?
2. Does this command changes owner for **all** files and directories inside $HADOOP_HOME/? 
chown root:hadoop $HADOOP_HOME/*

dahachm replied:
> Why didn't "chown -r" work?

*/bin/sh* doesn't support **-r** parameter. I just checked its "help page", it does support **-R** parameter :'D 
So I really had a chance to make it one-line comand 

> Does this command changes owner for all files and directories inside $HADOOP_HOME/?
chown root:hadoop $HADOOP_HOME/*

No, it apply changes only for **1-st layer** of $HADOOP_HOME dir. Using **-R** would do the job, but there was no need for that :)

Hadoop was complaing on having lack of permissions for writing in home dir. Now I'm thinking that change owner of $HADOOP_HOME dir and $HADOOP_HOME/logs (after fomatting of FS) later  would be just enough

Shlepanets commented on 1-Docker/scripts/set-hosts-headnode.sh:
```diff
@@ -0,0 +1,14 @@
+
```
I really like that you've used awk here, Konstantin Radostev will be very proud of you =)
But jokes aside, what if, let's say, you would need to launch more than one worker in a cluster? Or if I decide to launch worker at first, and then headnode?
Is there a better way to communicate between containers?
Hint: you can already ping from one container to another, even without executing those set-hosts scripts, and without changing /etc/hosts. How?

dahachm replied:
Yes, you're right! There is much easier way to find out IP address of other nodes!

From docker documentation:
> By default Compose sets up a single network for your app. Each container for a service joins the default network and is both reachable by other containers on that network, and discoverable by them at a hostname identical to the container name.

So I can get IP address of other node by using **dig** (from bind-utils), for instance

Here I tried to resolve **headnode** name from **worker**, it works:
![Screenshot_5](https://user-images.githubusercontent.com/40645030/113356588-25be7080-934b-11eb-8e5f-e948aa24461b.png)

So there is no need to add smth to /etc/hosts

Shlepanets reviewed (commented):
General comments: 
- there is no need to expose that amount of ports. Usually, you should expose only necessary ports which will be accessed from "outside" network, such as Resource Manager UI (8088), NodeManager UI (8042), NameNode UI
- you should use "git squash" or similar command, in order to avoid merging 10 commits with the same "Update README.md"

dahachm commented:
@Shlepanets , thank you a lot!

I will consider all your comments and update ASAP!

dahachm committed:
Remove set-hosts scripts

	Containers are already able to resolve each other hostname.
	No need to add smth to /etc/hosts

dahachm committed:
Update docker-compose.yml

	Delete extra ports.
	Leave only 9870 (NameNode WebUI), 8088 (ResourceManager WebUI), 8042 (NodeManager WebUI)

	Update images name

dahachm committed:
Update README.md

dahachm commented:
@Shlepanets,

1. Removed *set-hosts* scripts and updated *start-*.sh* scripts (as there is no need to edit /etc/hosts)

2. Updated Docker Hub repo, now there are 2 images in different repos (https://hub.docker.com/repositories):
   **headnode**:  `docker pull dahachm/hadoop-headnode:1.0`
   **worker**:  `docker pull dahachm/hadoop-worker:1.0`
 
3. Updated *docker-compose.yml*:  deleted excess posrts

4. Rebased history of comits. I do remember, that it's a bad idea to do so after pushing in remote repository, but I will keep in mind  that all the rebase stuff must be done before pushing :)
